### PR TITLE
[quant][core] Add Embedding and EmbeddingBag reference module

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73436

Summary:
This PR adds support reference module support for Embedding and EmbeddingBag, following https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

* the reference module inherits from the corresponding float module (e.g. nn.Embedding), and the ReferenceQuantizedModule (which defines some utility functions to store qparms for a single weight)
* in forward, we first quantize and then dequantize weight (to generate the pattern) and then feed the weight to the original fp32 op

We'll connect this with fx grpah mode quantization later, in the final PR that deprecates the current convert implementation. Since current convert doesn't
support emitting quantize_per_tensor_dynamic ops, we don't want to implement it and immediately throw away the code, so might be better to just implement this
in the final flow.

Test Plan:
python test/test_quantization.py TestReferenceQuantizedModule.test_sparse
Integration test with the flow will be done later

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34480325](https://our.internmc.facebook.com/intern/diff/D34480325)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

**Review by vkuzo — COMMENTED** [review]
should we have a correctness test for the new modules?

**Comment by jerryzh168** [comment]
yes, planning to do that in a later PR that changes the current convert.py to something new, it will be a large PR so trying to land these things first to reduce the size. we may fix errors in that PR as well

**Comment by jerryzh168** [comment]
added some tests, please take a look again, thanks!

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

File: test/quantization/core/test_quantized_module.py [hunk 1]
```diff
@@ -1591,3 +1597,66 @@ def test_rnn(self):
...
             ref_res = ref_rnn(x, (h, c))
             self.assertEqual(fp32_res, ref_res)
+
+    def test_sparse(self):
+        """ Embedding and EmbeddingBag
+        """
+        num_embeddings = 10
+        embedding_dim = 3
+        # embedding input
+        ex = torch.LongTensor([[1,2,4,5],[4,3,2,9]])
```
Comment (vkuzo) [inline test/quantization/core/test_quantized_module.py:1607]: nit: lint

**Review by vkuzo — APPROVED** [review]

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34480325).

**Pull request closed without merging**